docs/Tutorial_DAMP.ipynb

```diff
@@ -0,0 +1,778 @@
+{
```

Comment:
<p>Why is this considered "offline/batch"? </p><p> </p><p>Separately, it feels too advanced to start off with splitting the data. You haven't really motivated why this splitting is necessary. Why not just take a time series, say that you are applying DAMP to all of it (without splitting) and then find the discord(s)?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/872/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Reply:
<p>[I will read the paper to provide the answers to your questions. For now, I am going to provide some explanations that come to my mind as I do not want to forget them later. I will keep this series of comments open]</p><p> </p><p>I think the only reason for the term "offline" here is that we use the look-ahead (<code>_forward_process</code> ) to prune some of the upcoming subsequences (so we know that those subsequences cannot be a discord and hence we do not need to do <code>_backward_process</code>   for them). As mentioned in the MATLAB code (and I think in the paper too), when look-ahead is set to 0, i.e. no pruning, then the algorithm becomes online. </p><p> </p><p> </p><blockquote>You haven't really motivated why this splitting is necessary. Why not just take a time series, say that you are applying DAMP to all of it (without splitting) and then find the discord(s)?</blockquote><p> </p><p>[Right! I need to go and check the paper to see what justification they have provided regarding this matter.]</p><p> </p><p>I think one of the reason is that we are looking LEFT discord! So, some of the first subsequences have no non-trivial LEFT neighbour. The very first subsequence that has one non-trivial neighbour cannot be good candidate as we haven't got enough information to whether choose that subsequence as discord or not. </p><p> </p><p>[I will elaborate more in the upcoming days]</p>

Reply:
<p><strong>Regarding "why </strong><code><strong>offline</strong></code><strong> ":</strong></p><p>As a follow up to my previous comment, below I have provided three bullet points mentioned in the DAMP paper regarding <code>lookahead</code> </p><p> </p><ul><li>If lookahead is zero, DAMP is a fast true online algorithm.</li><li>If lookahead is allowed to be arbitrarily large, DAMP is an ultrafast batch algorithm</li><li>Even if lookahead is a small (but non-zero) number, DAMP is effectively online algorithm,</li></ul><p> </p><p>The paper also says:</p><p> </p><blockquote>The algorithm as explained in Table 1 is a batch algorithm. To make it an online algorithm, we simply must reduce the size of the lookahead (Table 3, line 1) to the largest delay we are willing to accept (including possibly zero delay)</blockquote><p> </p><p>The current implementation in the notebook follows the algo provided in the Tables 1,2,3 of the paper.  The current implementation uses next-power-of-two of <code>m</code>  as the value of <code>lookahead</code> and thus, it is not small. Hence, it is offline algorithm. </p><p> </p><p> </p><p><strong>Regarding "why splitting is neccary":</strong></p><p>Personally I think the term train/test split, as used in the paper, may not make sense. The reason behind my opposition is that the train set (i.e. <code>T[:split_idx]</code> ) is not fixed  and it will change as we move towars to the end of time series. Note that for the query <code>T[query_idx : query_idx+m]</code> , we compare it with all of its left neighbors, i.e subsequences in <code>T[:query_idx]</code> , and not just the ones in <code>T[:split_idx]</code>   </p><p> </p><p>The paper also provides discussion on the <code>split_idx </code> :</p><blockquote>Recall that Classic DAMP has a parameter called spIndex, which sets the location of the split point between the training and test data in the initial state. When Classic DAMP processes a time series, it assumes that the data before spIndex, T[1:spIndex-1] are normal, which may lead to three issues. First, this causes the algorithm to ignore the potential anomalous behavior present in T[1:spIndex-1], resulting in certain false-negative results. Second, this approach may have the algorithm wasting time searching redundant data. It is possible that the patterns in T[1:spIndex-1] are highly redundant, such as 1,000 heartbeats that are essentially identical. If the heartbeats all have the same pattern, it would suffice for the algorithm to take just one of them to learn3 ; there is no need to consider the same pattern 1,000 times, which will waste a lot of time. Further, it may be difficult for T[1:spIndex-1] to contain every normal pattern, which can cause the algorithm to incorrectly identify normal behavior that does not appear in&nbsp; T[1:spIndex-1] as an anomaly. For example, if T[1:spIndex-1] only contains data on the solar zenith angle during the day, the algorithm may incorrectly identify normal solar zenith angles at night as anomalies. These potential problems can undermine the accuracy and efficiency of the algorithm</blockquote><p> </p><blockquote>Golden DAMP is our proposed solution to the above three problems</blockquote><p> </p><p>Two notes:</p><p>(1) As discussed above, <code>split_idx</code> is a timestamp before which ALL regular behaviors of the system appear with No anomaly. </p><p> </p><p>(2) As discussed above, the paper will handle the <code>split_idx</code> with the algorithm called Golden DAMP. In this algorithm, we have a vector (a time series) called Golden batch that contains ALL regular behaviors (without any redundant patters) . We can start from the very first index of the main input <code>T</code> . Although it has no left neighbors, That is not an issue sinec we do NOT look at the left neighbors anymore! We just compare a query with a Golden Batch. </p>

Reply:
<p>Please let know if further discussion is needed.</p>

Reply:
<p>I think we need to build up the intuition step-by-step. To me <code>split_idx</code> seems like an advanced topic and not a parameter that is automatically intuitive to the average user. How can we build up to using <code>split_idx</code>? </p>

Reply:
According to the paper, `split_idx` is problem dependent and user needs to have some insight about the system and its usual / regular patterns. So, they can say "upto the time stamp `split_idx`, the data shows [all kind of] regular behaviors and there is no anomaly". 

I added some explanation regarding this matter to the notebook.